spring-boot/proxy/src/test/java/ApiKeyAuthTokenVerifierTest.java

```diff
@@ -0,0 +1,185 @@
+        shallApiKeyBeValid(false);
+
+        assertNull(apiKeyAuthTokenVerifier.authenticate(apiKeyAuthToken));
+    }
+
+    @Test
+    public void nullApiKeyReceivedFromAdminShouldBeRejected() throws ConnectException {
+        setResultOfApiKeyRequestToAdmin(ResultOfApiKeyRequestToAdmin.NULL);
+        shallApiKeyBeValid(false);
+
+        assertNull(apiKeyAuthTokenVerifier.authenticate(apiKeyAuthToken));
+    }
+
+    @Test
+    public void apiKeyNotFoundInLocalApiKeyListShouldBeRejected() throws ConnectException {
+        setResultOfApiKeyRequestToAdmin(ResultOfApiKeyRequestToAdmin.EXCEPTION);
+        shallApiKeyBeFoundInLocalApiKeyList(false);
+
+        assertNull(apiKeyAuthTokenVerifier.authenticate(apiKeyAuthToken));
+    }
+
+    @Test
+    public void apiKeyAuthorizedByLocalApiKeyListAndShouldBeAuthorized() throws ConnectException {
+        setResultOfApiKeyRequestToAdmin(ResultOfApiKeyRequestToAdmin.EXCEPTION);
+        shallApiKeyBeFoundInLocalApiKeyList(true);
+        shallApiKeyBeAuthorizedConsideringLocalApiKeyList(true);
+
+        assertTrue(apiKeyAuthTokenVerifier.authenticate(apiKeyAuthToken).isAuthenticated());
+    }
+
+    @Test
+    public void apiKeyNotAuthorizedByLocalApiKeyListAndShouldNotBeAuthorized() throws ConnectException {
+        setResultOfApiKeyRequestToAdmin(ResultOfApiKeyRequestToAdmin.EXCEPTION);
+        shallApiKeyBeFoundInLocalApiKeyList(true);
+        shallApiKeyBeAuthorizedConsideringLocalApiKeyList(false);
+
+        assertNull(apiKeyAuthTokenVerifier.authenticate(apiKeyAuthToken));
+    }
+
+
+    //Behavior of the mock objects
+
+    //apiKeyFetcherMock
+    private void setResultOfApiKeyRequestToAdmin(ResultOfApiKeyRequestToAdmin resultOfApiKeyRequestToAdmin) throws ConnectException {
+        switch (resultOfApiKeyRequestToAdmin){
+            case NULL:
+                when(apiKeyFetcherMock.requestApiKeyFromAdmin(anyString())).thenReturn(null);
+                break;
+            case API_KEY:
+                when(apiKeyFetcherMock.requestApiKeyFromAdmin(apiKey.getKeyValue())).thenReturn(apiKey);
+                break;
+            case EXCEPTION:
+                when(apiKeyFetcherMock.requestApiKeyFromAdmin(anyString())).thenThrow(new ConnectException("test"));
+                break;
+        }
+    }
+
+    enum ResultOfApiKeyRequestToAdmin {
+        NULL, API_KEY, EXCEPTION
+    }
+
+    //localApiKeyVerifierMock
+    private void shallApiKeyBeValid(boolean isApiKeyBeValid) {
+        when(localApiKeyVerifierMock.isApiKeyValid(apiKey)).thenReturn(isApiKeyBeValid);
+    }
+
+    private void shallApiKeyBeAuthorizedConsideringLocalApiKeyList(boolean isAuthorized) {
+        when(localApiKeyVerifierMock.isApiKeyAuthorizedConsideringLocalApiKeyList(apiKey.getKeyValue()))
+                .thenReturn(isAuthorized);
+    }
+
+    //localApiKeyManagerMock
+    private void shallApiKeyBeFoundInLocalApiKeyList(boolean isFoundInLocalApiKeyList) {
+        if (isFoundInLocalApiKeyList)
```

Comment:
Maybe improve readability of whole method using this approach?
```
ApiKey expectedReturnValue = isFoundInLocalApiKeyList ? apiKey : null;
when(localApiKeyManagerMock.findApiKeyFromLocalApiKeyList(apiKey.getKeyValue())).thenReturn(expectedReturnValue);

```

Reply:
Done.